source/base/hdf5.cc

```diff
@@ -0,0 +1,1752 @@
+
+    Container data = internal::initialize_container<Container>(dimensions);
+
+    if (mpi)
+      {
+        plist = H5Pcreate(H5P_DATASET_XFER);
+        Assert(plist >= 0, ExcInternalError());
+        ret = H5Pset_dxpl_mpio(plist, H5FD_MPIO_COLLECTIVE);
+        Assert(ret >= 0, ExcInternalError());
+      }
+    else
+      {
+        plist = H5P_DEFAULT;
+      }
+
+    ret = H5Dread(*hdf5_reference,
+                  *t_type,
+                  H5S_ALL,
+                  H5S_ALL,
+                  plist,
+                  internal::get_container_pointer(data));
+    Assert(ret >= 0, ExcInternalError());
+
+    if (mpi)
+      {
+        if (query_io_mode)
+          {
+            ret = H5Pget_mpio_actual_io_mode(plist, &io_mode);
+            Assert(ret >= 0, ExcInternalError());
+            ret = H5Pget_mpio_no_collective_cause(plist,
+                                                  &local_no_collective_cause,
+                                                  &global_no_collective_cause);
+            Assert(ret >= 0, ExcInternalError());
+          }
+        ret = H5Pclose(plist);
+        Assert(ret >= 0, ExcInternalError());
+      }
+
+    (void)ret;
+    return data;
+  }
+
+
+
+  template <typename Container>
+  Container
+  DataSet::read_selection(const std::vector<hsize_t> &coordinates)
+  {
+    Assert(coordinates.size() % rank == 0,
+           ExcMessage(
+             "The dimension of coordinates has to be divisible by the rank"));
+    const std::shared_ptr<hid_t> t_type =
+      internal::get_hdf5_datatype<typename Container::value_type>();
+    hid_t  plist;
+    hid_t  memory_dataspace;
+    herr_t ret;
+
+    std::vector<hsize_t> data_dimensions{
+      static_cast<hsize_t>(coordinates.size() / rank)};
+
+    Container data = internal::initialize_container<Container>(data_dimensions);
+
+    memory_dataspace = H5Screate_simple(1, data_dimensions.data(), NULL);
+    Assert(memory_dataspace >= 0, ExcMessage("Error at H5Screate_simple"));
+    ret = H5Sselect_elements(*dataspace,
+                             H5S_SELECT_SET,
+                             data.size(),
+                             coordinates.data());
+    Assert(ret >= 0, ExcMessage("Error at H5Sselect_elements"));
+
+    if (mpi)
+      {
+        plist = H5Pcreate(H5P_DATASET_XFER);
+        Assert(plist >= 0, ExcMessage("Error at H5Pcreate"));
+        ret = H5Pset_dxpl_mpio(plist, H5FD_MPIO_COLLECTIVE);
+        Assert(ret >= 0, ExcMessage("Error at H5Pset_dxpl_mpio"));
+      }
+    else
+      {
+        plist = H5P_DEFAULT;
+      }
+
+    ret = H5Dread(*hdf5_reference,
+                  *t_type,
+                  memory_dataspace,
+                  *dataspace,
+                  plist,
+                  internal::get_container_pointer(data));
+    Assert(ret >= 0, ExcMessage("Error at H5Dread"));
+
+    if (mpi)
+      {
+        if (query_io_mode)
+          {
+            ret = H5Pget_mpio_actual_io_mode(plist, &io_mode);
```

Comment:
The functions `read_*` and `write_*` of `DataSet` "could" be const. They are not const because I would like to give the option to do some additional checking and store the information about the collective call somewhere. Most of the people won't use `query_io_mode`, only those who really care about high HDF5 performance will use it.

I see several options. Let me know if you have a better idea.

### Current option - option A
- I prefer this option because the interface is simple
- If `query_io_mode = True`, the functions `read_*` and `write_*` will do additional checking and write information about the collective operations in the object `Dataset`.
- read_* and write_* can't be const because `DataSet` is modified
- The advantage of this option is that the additional overhead can easily be turned off with `set_query_io_mode(False)`.
``` c++
#ifdef DEBUG
dataset.set_query_io_mode(true);
#endif
```

### Option B
- read_* and write_* can be const
- The interface becomes complicated
- In the user code the Release vs Debug version becomes more complicated.
``` c++
#ifdef DEBUG
write(data, io_mode, local_no_collective_cause, global_no_collective_cause) ;
#else
write(data) ;
#endif
```
- I can create 3 overloads of read_* and write_*. One does a simple read/write and the two other do the additional checking:
``` c++
void write(const Container &data) const;
void write(const Container &data, H5D_mpio_actual_io_mode_t &io_mode, H5D_mpio_actual_io_mode_t &local_no_collective_cause, H5D_mpio_actual_io_mode_t &global_no_collective_cause) const;
void write(const Container &data,std:: string &io_mode, std::string &local_no_collective_cause, std::string &global_no_collective_cause) const;
```

Reply:
Option A is fine. I was just wondering.

In hindsight, I think that's the right choice anyway. It's true that maybe in `ostream::write()` or `istream::read()` the *file handle* doesn't change, but the *state* of the file surely does. Consequently, it seems correct, at least *semantically*, for these functions to not be `const`.